### DISCUSSION TITLE
The ScalarBar, OrientationAxes, and AxesGrid fail to render when calling paraview.VtkLocalView

### DISCUSSION THREAD

**Discussion by smkacur** [Kitware/trame#524] [Trame/ParaView] [answered]
Hello, I have a trame/paraview application written in vue3. When I test the app locally using VtkRemoteView, everything works as expected. But when I swtich to VtkLocalView, which I need for deployment, only the VTKs render but none of the other elements. 

Do you have any thoughts as to why the other elements are lost in LocalView?

In case it helps, here is the last part of the GUI
```python
    with layout.content:
        with vuetify.VContainer(fluid=True, classes="fill-height pa-0 ma-0"):
            html_view = paraview.VtkLocalView(
                view, namespace="view")
            ctrl.view_update = html_view.update
            ctrl.view_reset_camera = html_view.reset_camera
```

And here is the callback where I define the scalarbar

```python
@ state.change("active_array")
def update_color_by(active_array, fields, **kwargs): 
    global color_bar

    if len(fields) == 0:
        return

    array = fields[active_array]
    simple.ColorBy(representation, (array.get("location"), array.get("title")))
    representation.RescaleTransferFunctionToDataRange(True, False)
    name = pv.make_name_valid(array.get("title"))
    lut = simple.GetColorTransferFunction(name)
    pwf = simple.GetOpacityTransferFunction(name)
    _min, _max = array.get("range")
    lut.RescaleTransferFunction(_min, _max)
    pwf.RescaleTransferFunction(_min, _max)
    color_bar = simple.GetScalarBar(lut, view)
    color_bar.Title = array.get("title")
    color_bar.RangeLabelFormat = '%-#3.2f'
    color_bar.ComponentTitle = ""
    color_bar.WindowLocation = 'Lower Center'
    color_bar.Orientation = 'Horizontal'
    simple.UpdateScalarBars(view)
    ctrl.view_update()
```

**Comment by jourdain** [comment, marked as answer]
Those discrepancy are expected when using vtk.js as rendering implementation. But we are working on a WASM implementation that will fix that. But it won't be ready for prime time before 5.13 at least.

**Reply by smkacur** [reply]
Can you please point me to a place where I can follow the progress?

**Comment by jourdain** [comment]
You can get most of it working if you create the scalar bar in plain VTK and put it into the ParaView view.

**Reply by smkacur** [reply]
Thanks for your response! I haven't worked with plain VTK before but I tried this to get the scalar bar working without rewriting the entire function in vtk:

```python
# Added to imports
from vtkmodules.vtkRenderingAnnotation import vtkScalarBarActor
from vtkmodules.vtkRenderingCore import vtkRenderer

# Added below section where I define the view with simple.GetRenderView
vtk_renderer = vtkRenderer()
vtk_renderer.AddActor2D(color_bar)  # color_bar is a global_variable

# Changed the update_color_by function
@ state.change("active_array")
def update_color_by(active_array, fields, **kwargs):  # viewMode="remote",
    global color_bar

    if len(fields) == 0:
        return

    array = fields[active_array]
    simple.ColorBy(representation, (array.get("location"), array.get("title")))
    representation.RescaleTransferFunctionToDataRange(True, False)
    name = pv.make_name_valid(array.get("title"))
    lut = simple.GetColorTransferFunction(name)
    pwf = simple.GetOpacityTransferFunction(name)
    _min, _max = array.get("range")
    lut.RescaleTransferFunction(_min, _max)
    pwf.RescaleTransferFunction(_min, _max)

    color_bar = vtkScalarBarActor()
    color_bar.SetLookupTable(lut.GetClientSideObject())
    color_bar.SetTitle(array.get("title"))
    color_bar.SetLabelFormat('%-#3.2f')
    color_bar.SetComponentTitle("")
    color_bar.SetOrientationToHorizontal()
    color_bar.RenderOverlay(vtk_renderer)
    ctrl.view_update()
```
    
When I call `color_bar.RenderOverlay(vtk_renderer)` I get a segmentation fault. If I don't call RenderOverlay then I don't get any errors but the color_bar doesn't render. How would you suggest adding the vtkScalarBarActor() to the paraview view?

**Reply by jourdain** [reply]
Something like `simple.GetRenderView().GetRenderer().AddActor2D(color_bar)`

**Reply by smkacur** [reply]
Thank you that works well!

**Reply by jourdain** [reply]
The WASM on ParaView has not started, but you can follow the trame side with VTK [here](https://github.com/Kitware/trame-vtklocal).

**Reply by jourdain** [reply]
The main work will be in ParaView repo in C++/CMake.

**Comment by smkacur** [comment]
Forgive the stupid question but if I don't have a GPU in my server is it still possible to use paraview.VtkRemoteRendering?

Maybe a better question is when should I use Local vs Remote rendering?

**Reply by jourdain** [reply]
See [doc](https://kitware.github.io/trame/guide/tutorial/vtk.html#local-and-remote-rendering) for Local vs Remote.

Otherwise you can use remote without GPU but it will be slow. You just need to use the egl or osmesa version of ParaView.

**Reply by smkacur** [reply]
Thank you.
